### PR TITLE
cups-drivers: Brother HL-L3230CDW driver version 1.0.2-0

### PR DESCRIPTION
**Pull request by aplund**
###### Description of changes

Add cups drivers for Brother HL-L3230CDW

###### Things done

- Built on platform(s)
  - [ x ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+in
+stdenv.mkDerivation {
+
+  pname = "cups-brother-hll3230cdw";
+  inherit version;
+
+  buildInputs = [ cups perl ];
+  nativeBuildInputs = [ dpkg ];
+
+  inherit src;
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/hll3230cdw/default.nix:19]:
```suggestion
stdenv.mkDerivation {
  pname = "cups-brother-hll3230cdw";
  version = "1.0.2";
  
  src = fetchurl {
    url = "https://download.brother.com/welcome/dlf103925/hll3230cdwpdrv-${version}-0.i386.deb";
    sha256 = "9d49abc584bf22bc381510618a34107ead6ab14562b51831fefd6009947aa5a9";
  };

  nativeBuildInputs = [ dpkg ];

  buildInputs = [ cups perl ];
```

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  dontUnpack = true;
+
+  installPhase = ''
+    dpkg-deb -x "$src" "$out"
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/hll3230cdw/default.nix:23]: unpacking should always be done in unpackPhase. Please see unpackCmd, too.

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    mkdir -p "$out/usr/lib/cups/filter"
+    ln -s "$out/opt/brother/Printers/hll3230cdw/cupswrapper/brother_lpdwrapper_hll3230cdw" \
+      "$out/usr/lib/cups/filter/brother_lpdwrapper_hll3230cdw"
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/hll3230cdw/default.nix:26]: Do we need to use usr/lib? Can't we just use lib? Also why do we need opt? Can we just copy the file into the final location straight away?

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  meta = with lib; {
+    description = "Brother HL-L3230CDW printer driver";
+    license = licenses.unfree;
+    sourceProvenance = with sourceTypes; [ binaryNativeCode ];
+    platforms = platforms.linux;
+    homepage = "http://www.brother.com/";
+    downloadPage = "https://support.brother.com/g/b/downloadend.aspx?c=us&lang=en&prod=hll3230cdw_us_eu_as&os=128&dlid=dlf103925_000&flang=4&type3=10283";
+  };
+
+}
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/hll3230cdw/default.nix:41]:
```suggestion
  };
}
```

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      "$out/share/cups/model/brother_hll3230cdw_printer_en.ppd"
+  '';
+
+  meta = with lib; {
+    description = "Brother HL-L3230CDW printer driver";
+    license = licenses.unfree;
+    sourceProvenance = with sourceTypes; [ binaryNativeCode ];
+    platforms = platforms.linux;
+    homepage = "http://www.brother.com/";
+    downloadPage = "https://support.brother.com/g/b/downloadend.aspx?c=us&lang=en&prod=hll3230cdw_us_eu_as&os=128&dlid=dlf103925_000&flang=4&type3=10283";
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/hll3230cdw/default.nix:38]: maintainers is missing

**Review by SuperSandro2000 — COMMENTED** [review]
You are not referencing the file anywhere so the package won't be build.

Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix (hunk 3 above)
Comment (aplund) [reply pkgs/misc/cups/drivers/hll3230cdw/default.nix]: I can't figure out a good solution here.  The distributed files in /opt are in a bit of a strange arrangement which I can't figure out.  I basically did the same thing as the other brother CUPS drivers.

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix (hunk 5 above)
Comment (aplund) [reply pkgs/misc/cups/drivers/hll3230cdw/default.nix:41]: This has been added in the updated branch.

**Comment by aplund** [comment]
The brother scripts and binaries use the location of the files in /opt/brother to figure out parameters like the name of the printer.  It isn't really possible to move them around.  I've tried to take approaches in other brother drivers to compensate for this as much as possible.  I'm still a little confused as to what's going on with the linker as the fixup doesn't seem to change the interpreter parameter in the ELFs.

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  # The configuration binary 'brprintconf_hll3230cdw' and lpd filter
+  # 'brhll3230cdwfilter' has hardcoded /opt format strings.  There isn't
+  # sufficient space in the binaries to substitute a path in the store, so use
+  # libredirect to get it to see the correct path.  The configuration binary
+  # also uses this format string to print configuration locations.  Here the
+  # wrapper output is processed to point into the correct location in the
+  # store.
+
+  postFixup = ''
+
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/hll3230cdw/default.nix:51]:
```suggestion
```

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    mkdir -p "$out/lib/cups/filter"
+    mkdir -p "$out/share/cups/model"
+
+
+    ln -s "$out/opt/brother/Printers/hll3230cdw/cupswrapper/brother_lpdwrapper_hll3230cdw" \
+      "$out/lib/cups/filter/brother_lpdwrapper_hll3230cdw"
+    ln -s "$out/opt/brother/Printers/hll3230cdw/cupswrapper/brother_hll3230cdw_printer_en.ppd" \
+      "$out/share/cups/model/brother_hll3230cdw_printer_en.ppd"
+
+
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/hll3230cdw/default.nix:35]:
```suggestion
```

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    runHook preInstall
+
+    mkdir -p "$out"
+    cp -pr opt "$out"
+    cp -pr usr/bin "$out/bin"
+    rm "$out/opt/brother/Printers/hll3230cdw/cupswrapper/cupswrapperhll3230cdw"
+
+    mkdir -p "$out/lib/cups/filter"
+    mkdir -p "$out/share/cups/model"
+
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/hll3230cdw/default.nix:28]:
```suggestion
    mkdir -p "$out/lib/cups/filter" "$out/share/cups/model"
```

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix [hunk 9]
```diff
@@ -0,0 +1,81 @@
+{ stdenv, lib, fetchurl, perl, gnused, dpkg, makeWrapper, autoPatchelfHook, libredirect }:
+
+stdenv.mkDerivation rec {
+
```
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/hll3230cdw/default.nix:4]:
```suggestion
```

File: pkgs/misc/cups/drivers/hll3230cdw/default.nix (hunk 3 above)
Comment (aplund) [reply pkgs/misc/cups/drivers/hll3230cdw/default.nix]: I have borrowed from other brother cups driver pkgs to deal with this issue.  In their infinite wisdom, brother uses the path in /opt to get information about which driver it is using and then to load the associated configuration.  I have patched most of this insanity like other packages, but I don't believe one can just rearrange the paths as the paths into /opt are hardcoded into resources of the binaries.  The solution involve using libredirect for file operations and sed on the output to always point into the store.

**Pull request merged**